File: implementations/cpp/libwebp/decode.cpp
#include <webp/decode.h>

#include <stdexcept>
#include <vector>

#include "benchmark_harness.hpp"

class LibWebpBench : public BenchmarkImplementation {
 public:
  std::string name() const override { return "libwebp-decode"; }

  void prepare(const Args &args) override {
    input_data = read_binary_file(args.input);
  }

  std::vector<uint8_t> run(const Args &args) override {
    return decode(input_data);
  }

 private:
  std::vector<uint8_t> decode(const std::vector<uint8_t> &data) {
    int width, height;
    uint8_t *output_buffer =
        WebPDecodeRGB(data.data(), data.size(), &width, &height);

    if (output_buffer == nullptr) {
      throw std::runtime_error("WebPDecodeRGB failed");
    }

    std::vector<uint8_t> output(output_buffer,
                                output_buffer + width * height * 3);
    WebPFree(output_buffer);

    return encode_ppm_rgb8(width, height, output);
  }

  std::vector<uint8_t> input_data;
};

int main(int argc, char **argv) {
  LibWebpBench bench;
  return run_benchmark(argc, argv, bench);
}

File: implementations/rust/image-webp/src/decode.rs
use anyhow::{Context, Result};
use benchmark_harness::{Args, BenchmarkImplementation};
use image::ImageFormat;
use std::fs;

struct ImageWebpBench;

struct BenchContext {
    input_data: Vec<u8>,
}

impl BenchmarkImplementation for ImageWebpBench {
    fn name(&self) -> &'static str {
        "image-webp-decode"
    }

    fn prepare(&self, args: &Args) -> Result<Box<dyn std::any::Any>> {
        let input_data = fs::read(&args.input).context("Failed to read input file")?;

        Ok(Box::new(BenchContext { input_data }))
    }

    fn run(&self, _args: &Args, context: &mut dyn std::any::Any) -> Result<Vec<u8>> {
        let ctx = context
            .downcast_ref::<BenchContext>()
            .expect("Invalid context");
        let img = image::load_from_memory_with_format(&ctx.input_data, ImageFormat::WebP)
            .context("Failed to decode WebP")?;

        // Output as PPM
        let rgb = img.to_rgb8();
        let mut output = Vec::with_capacity(15 + rgb.len()); // Header + Data
        use std::io::Write;
        write!(&mut output, "P6\n{} {}\n255\n", rgb.width(), rgb.height())?;
        output.write_all(&rgb)?;

        Ok(output)
    }
}

fn main() -> Result<()> {
    benchmark_harness::main(ImageWebpBench)
}
